### ISSUE TITLE
FlushLightStepTracer is not reliable

### ISSUE THREAD

**Issue by RaduBerinde** [lightstep/lightstep-tracer-go#89]
We have a situation where we are tracing some short-lived tests and some traces aren't making it to lightstep. We added a call to `FlushLightStepTracer` and that seems to help, but it still doesn't seem reliable. Sometimes we see this error:
```
I170621 15:52:43.061339 50 util.go:39  LightStep instrumentation error (a previous Report is still in flight; aborting Flush()). Set the Verbose option to enable more logging.
```

The flush code indeed just bails if a `Report` is in progress. This is problematic on two fronts:
 - we won't flush any traces that were generated since we started this last Report
 - more importantly - normally, the `FlushLightStepTracer` call returns only after we've flushed, so it's reasonable to call this before exiting. But if we hit this error, we won't even wait for the in-progress Report to finish.

I think that this case should be handled by the flush function - it should wait until the current Report finishes and then start another one.

**Comment by RaduBerinde** [comment]
A possible partial "fix" would be to at least return the error from `Flush()` so the caller can do a retry loop.

**Comment by tedsuo** [comment]
Hi @RaduBerinde I see the case you are discussing - flush should wait for the current report to complete, and possibly send another report if there is one in flight.

We can also add an error in the case where the collector connection failing. However, it is unclear to me that we should encourage retries on error - any retry logic that has a hope of being fruitful should be internal to the tracer. Also, FYI we are adding a centralized error handler for reporting purposes: https://github.com/lightstep/lightstep-tracer-go/issues/91

**Comment by RaduBerinde** [comment]
I agree, it would be preferable to not rely on the caller retrying.

**Comment by RaduBerinde** [comment]
By the way, I think `CloseTracer` has a similar problem - I posted about it in this PR (unfortunately, a while after it was merged): https://github.com/lightstep/lightstep-tracer-go/pull/85#discussion_r123800322

**Comment by tedsuo** [comment]
Sorry, I missed your comments on the previous PR! But I agree with you - Close has the same problem. Also, we have internal errors with no reporting mechanism. So I think we can solve these issues together in v0.14:

- Centralize error handling with `OnError` handler
- `Flush` ensures messages in buffer prior to invocation are reported.
- `Close` ensures buffer is empty before returning.
- `Close` and `Flush` can be aborted via `context.Context`

```go
import ot "github.com/opentracing/opentracing-go"

// Backwards compatible convenience methods. The only error returned 
// here is from failing to typecast to a lighstep Tracer.
func FlushLightStepTracer(ot.Tracer, ...context.Context) error {}
func CloseTracer(ot.Tracer, ...context.Context) error {}

// Options takes an error handler
type Options struct {
    OnError func(error)
}

// Close and Flush take a context, lose an error
type Tracer interface {
    ot.Tracer
    Close(context.Context) 
    Flush(context.Context)
    Options() Options
}
```

**Comment by ChimeraCoder** [comment]
+1 on this

`FlushLightStepTracer` should return an error from the flushing operation - right now, [it always returns nil](https://github.com/lightstep/lightstep-tracer-go/blob/master/tracer.go#L24), so checking the error value doesn't yield anything.

Ideally, I'd strongly prefer having the error there as opposed to delivered asynchronously, because that's the point where it makes the most logical sense for us to handle it.

That said, when we finish a span using `FinishWithOptions`, there's no way to ensure that we're getting the error, if it could have been flushed in the background before the call to `FlushLightstepTracer`. If there's a way to finish a span *without* flushing (ie, a way to control flush operations and ensure that they can always happen in the foreground), that would be ideal. Barring that, asynchronous error delivery could work, though it'd be clunkier.

Either way, we'd definitely need a way to monitor this. Right now, if a collector disappears for any reason, all traces sent to it will be silently dropped, and the only way to prevent that is to continuously redial the collector to guarantee reaching a fresh one.

**Comment by tedsuo** [comment]
@ChimeraCoder thanks for the feedback.

In response to handling errors synchronously: we feel like there should be no situation that warrants this. For flush, we would prefer to handle backoff/retry internally, so that there are only three scenarios when calling flush:

a) the report is sent successfully.
b) the tracer determines that flushing is impossible, retry will not work.
c) the caller cancels the attempt.

The main reason for having an error on return of flush (instead of a context for cancelation) is to attempt retry/backoff logic outside of the tracer. But backoff should only occur at one layer: both the tracer and the client should not be implementing backoff. As the tracer is in a better position to implement an effective backoff strategy, the client should simply wait, aborting if a deadline is exceeded.

A centralized error handler allows for error reporting, along with the possibility of exiting the program or swapping out the global tracer with a new one. All of those operations should be handled centrally; we’d like to eliminate any other need for error handling.

(In general, the expectation in OpenTracing is that code will not abide any interference from instrumentation, and instrumentation blackouts are not a sufficient reason to induce latency or failure into the underlying operations. That is why OpenTracing does not have error returns for span operations.)

Please let me know if that sounds like a good plan to you. :)

Cheers,
Ted

**Comment by tedsuo** [comment]
In v0.15, a `context` parameter was added to `Flush` and `Close` to help manage shutdown. Also, the `error` return value was removed to avoid confusion. If a non-lightstep tracer is passed to the static helper functions, `EventUnsupportedTracer` will be emitted instead of an error being returned.

https://github.com/lightstep/lightstep-tracer-go#flushing-and-closing-managing-the-tracer-lifecycle
https://godoc.org/github.com/lightstep/lightstep-tracer-go#EventUnsupportedTracer

We don't recommend wrapping `Flush` and `Close` in retry/backoff functions, please assume these functions will block until successful or retry has failed.

I believe this change resolves this issue. If you have any further trouble with how `Flush` or `Close` have been implemented, please reopen or file a new issue.

Cheers,
Ted
